content/post/performance-metrics-histogram.md

```diff
@@ -0,0 +1,245 @@
+
+__**Metrics Receiver**__  
+A instance on each Wallaroo worker that is responsible for collecting and processing metrics produces on the worker. The same instance is also responsible for emitting the aggregate collection to an external monitoring system.
+
+## Problem Statement
+
+We would like to provide our clients with an SLA stating that 99% of the events we process have an end-to-end latency of less than 1ms, and 99.9% have an end-to-end latency of less than 1s.
+We wish to be able to expose this information for an arbitrary, continuous time range. Our current monitoring system shows the data for a sliding 5-minute window, which is updated once per second.
+In addition, we would also like to see how well our application is performing overall so we can identify areas for improvement and areas where the application is already competitive and are worth highlighting to potential customers.
+
+To summarize: what we are trying to achieve is the ability to tell whether 99% or more of the events processed in a given time period have a latency below 1ms, and whether 99.9% or more of the events processed in a given time period have a latency below 1s. In addition, we need to be able to efficiently recompute this every second.
+
+## Approaches
+
+## Using A Sorted Values List
+
+Using Sorted Values List, the following method is used:
+
+1. We store all data as a list of events, L. L is sorted by time.
+2. We then select a window, W, from L, such that W contains all events whose time is within the last 5 minutes  
+    `{W: e ∀ e ∊ L | e.time > now()-'5 minutes'}`
+3. To select the values in the 99% and 99.9% positions, we need to sort W. We call the sorted list W'.
+4. We then select the 99% and 99.9% positions in the sorted list W', by taking its length and multiplying it by 0.99 and 0.999 respectively, then rounding the fraction up to the nearest integer. We use those integers as the indexes for the events whose latencies are the 99%- and 99.9%-percentiles. 
+
+### Cost Analysis
+
+- We have to store every single event in L.
+    - We have to send every single event to the monitoring hub so it can store it in L.  
+        __**Space cost: O(|L|)**__
+- We have to index L by time. 
+    - We can assume events arrive in order of completion, so L is naturally sorted.  
+      __**Time cost: 0**__
+- We have to select the window W from L.  
+    __**Space cost: O(|W|)**__  
+    __**Time cost: O(log(|W|))**__  
+- We have to sort W in order of latency values.
+    - Assuming best case performance of quicksort of O(nlogn)  
+        __**Time cost: O(|W|log(|W|))**__  
+
+__**Total time cost: O(|W|log(|W|)) + O(log|W|)**__  
+__**Total space cost: O(|L|) + O(|W|)**__
+
+## Using a Histogram
+
+Using a histogram, the following method is used: 
+
+1. We aggregate the histograms from a stream of events.
```

Comment:
It seems this is missing a step describing events being captured/added into the histogram bins. Or maybe that's what substep `1.1` is supposed to be explaining? In either case, it would be helpful to have a sentence explaining that the latency for every event is stored in the histogram in the appropriate bin but the event itself is not stored to avoid any potential confusion by folks who may not be familiar with how histograms work.

Reply:
This is what subsection 1.1 was saying, but I'll reword it so it's clearer! Thanks for catching it.